.github/workflows/go.yml

```diff
@@ -1,56 +1,186 @@
+
+            - name: upload cover profile artifact
+              uses: actions/upload-artifact@v4
+              with:
+                name: coverage.out
+                path: coverage.out
+                if-no-files-found: error
+
+    check-coverage:
+        needs: test
+        runs-on: ubuntu-latest
+        name: Check Coverage
+        steps:
+            - name: checkout
+              uses: actions/checkout@v4
+
+            - name: Download cover profile artifact
+              id: download-coverage
+              uses: actions/download-artifact@v4
+              with:
+                name: coverage.out
+
+            - name: Extract coverage percentage
+              id: current-coverage
+              run: |
+                if [ -f coverage.out ]; then
+                  COVERAGE=$(go tool cover -func=coverage.out | grep total: | awk '{print $3}' | sed 's/%//')
+                  echo "coverage=$COVERAGE" >> $GITHUB_OUTPUT
+                else
+                  echo "coverage=0" >> $GITHUB_OUTPUT
+                fi
+            
+            - name: download artifact (master.breakdown)
+              id: download-master-breakdown
+              uses: dawidd6/action-download-artifact@v9
+              with:
+                branch: master
+                workflow_conclusion: success
+                name: master.breakdown
+                if_no_artifact_found: warn
+
+            - name: download artifact (master-coverage.out)
+              id: download-master-coverage
+              uses: dawidd6/action-download-artifact@v9
+              with:
+                branch: master
+                workflow_conclusion: success
+                name: master-coverage.out
+                if_no_artifact_found: warn
+
+            - name: Extract master coverage percentage
+              id: master-coverage
+              run: |
+                if [ -f master-coverage.out ]; then
+                  MASTER_COVERAGE=$(go tool cover -func=master-coverage.out | grep total: | awk '{print $3}' | sed 's/%//')
+                  echo "coverage=$MASTER_COVERAGE" >> $GITHUB_OUTPUT
+                else
+                  echo "coverage=0" >> $GITHUB_OUTPUT
+                fi
+
+            - name: Generate full coverage breakdown
+              id: full_coverage_report
+              run: |
+                if [ -f coverage.out ]; then
+                  REPORT_CONTENT=$(go tool cover -func=coverage.out) # This command outputs function-level coverage [5]
+                  echo "report<<EOF" >> $GITHUB_OUTPUT # Start HERE-doc for multi-line output [3]
+                  echo "$REPORT_CONTENT" >> $GITHUB_OUTPUT
+                  echo "EOF" >> $GITHUB_OUTPUT # End HERE-doc
+                else
+                  echo "report=No coverage report found." >> $GITHUB_OUTPUT
+                fi
+
+            - name: check test coverage
+              id: coverage
+              uses: vladopajic/go-test-coverage@v2
```

Comment:
should we consider using codecov https://github.com/envoyproxy/ai-gateway/pull/759/files